docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+
+The first word contains the region metadata.
+This is a discriminated union where the low three bits describe how the remainder is used (for example, as a reference count for one immutable object in an SCC, a pointer to that object for other immutable objects in an SCC, and a pointer to the region metadata for sentinel objects).
+
+The second is a pointer to the descriptor for the object.
+The low bits of this are used by the runtime for memory management state and so must be zeroed before accesses.
+
+The object descriptor layout is still likely to change, the current layout is:
+
+64-bit integer size.
+Pointer to a trace function, used in traced regions to walk reachable objects.
+Pointer to a finalise function, called when an object is dead but before other objects in the same region are deleted (though they may have already been finalised.
+Pointer to a notify function, invoked when an object is notified of a noticeboard event.
+Pointer to a destructor, which is called after some objects in a region have been deallocated and may not follow any `mut` pointers.
+The remainder of the descriptor is the vtable.
+
+Methods in the vtable use *selector colouring*.
+Each selector (method name and types) is assigned a separate vtable index.
+Pairs of selectors that do not appear in the same interface are then combined to use the same index.
+Selector colouring is not essential but can dramatically reduce the size of vtables.
+
+*Note:* If we can distinguish between sentinel objects and others at object-creation time, the region metadata field for non-sentinel objects could be a 32-bit value that is either an integer or an index into a table, rather than a pointer.
+This would be a noticeable saving on CHERI systems and a small saving on 64-bit systems, at the cost of an extra indirection for some memory-management operations.
+We could also use a 32-bit index into a descriptor table, as the Fuchsia C++ ABI does: Programs that have more than 2^32 distinct types arise only as a result of poor life choices.
+
+*Note:* The above choices are visible to the runtime but are part of the internal ABI and so it would be possible to have size and performance optimised variants available on a per-program basis.
+
+*Note:* On AArch64 systems where the top-byte-ignore feature is available, using the top bits in the descriptor pointer instead of the bottom ones would improve efficiency of descriptor accesses.
+
+### Field layout
+
+In C, `struct` layout is a trade between storage density and accidental aliasing.
+If two fields are accessed concurrently, false sharing in the cache can cause significant performance bottlenecks.
+In Verona; however, no mutable objects are ever shared and so, aside from reference count manipulations on immutable objects and interactions between adjacently allocated objects in different regions, there are no negative interactions with modern caches.
+The Verona field layout algorithm is therefore free to optimise for space.
```

Comment:
Dynamically loadable code is an entirely separate discussion and is explicitly not planned for MVP, but I don't think this is actually a concern.  As long as the layout algorithm is the same in both cases, it will give the results.  This is not an optional compiler optimisation, this is the goal of the design of the layout algorithm.

Reply:
Right, I meant an older version of the compiler created `libfoo.so` and a newer version (with improved packing) created the executable `foo.exe`. But I guess this is just like any other ABI breaking change, people will have to recompile everything.

Means we'll have to track any change to this algorithm very closely to make sure there are no ABI-breaking scenarios.

We'll also need some ABI versioning in the objects to warn users that include older libraries into their code, but that's a separate discussion.

Reply:
Perhaps also make it explicit in this document that this encoding is a core part of the ABI, and that not only it's deterministic on the types only, but it has strong ABI guarantees and that we'll change the ABI version if any of that changes.

Reply:
I think you are working under the assumption of a C linkage model.  This is something that we have explicitly discounted for Verona.  Our loadable code model is not going to be supported at all in v1 and will not look like a UNIX .so when it is implemented.